### PR TITLE
Fixes #9699: UI to define datasources 

### PR DESCRIPTION
**Pull request by RaphaelGauthier**
https://www.rudder-project.org/redmine/issues/9699

### CODE REVIEW

File: rudder-core/src/main/scala/com/normation/rudder/datasources/DataSourceRepository.scala [hunk 1]
```diff
@@ -96,7 +96,11 @@ trait DataSourceUpdateCallbacks {
 
 class MemoryDataSourceRepository extends DataSourceRepository {
 
-  private[this] var sources : Map[DataSourceId,DataSource] = Map()
```
Comment (VinceMacBuche) [inline rudder-core/src/main/scala/com/normation/rudder/datasources/DataSourceRepository.scala]: Can you remove this ?

File: rudder-web/src/main/webapp/javascript/rudder/angular/dataSource.js [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+app.factory('focus', function ($rootScope, $timeout) {
+  return function(name) {
+    $timeout(function (){
+      $rootScope.$broadcast('focusOn', name);
+    });
+  }
+});
+
+app.controller("datasourceCtrl", ['$scope', '$timeout', 'orderByFilter','$http', function($scope,$timeout,orderBy,$http,focus) {
+  $scope.forms = {};
```
Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/angular/dataSource.js]: IS this variable set ?

File: rudder-web/src/main/webapp/javascript/rudder/angular/dataSource.js [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+  $scope.datasources = [];
+  // Selected data source
+  $scope.selectedDatasource;
+
+  /* Get data sources */
+  $scope.getDataSources = function(){
+    return $http.get(contextPath + '/secure/api/latest/datasources').then(function(response){
+      var res = response.data.data.datasources;
+      $scope.propertyName = 'name';
```
Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/angular/dataSource.js]: Why do you need this variable to be stored in $scope ?

File: rudder-web/src/main/webapp/javascript/rudder/angular/dataSource.js [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+      $scope.reverse = false;
```
Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/angular/dataSource.js]: same here

File: rudder-web/src/main/webapp/javascript/rudder/angular/dataSource.js [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+          'schedule'       : timeConvert($scope.datasources[i].runParameters.schedule.duration)
+        , 'updateTimeout'  : timeConvert($scope.datasources[i].updateTimeout)
+        , 'requestTimeout' : timeConvert($scope.datasources[i].type.parameters.requestTimeout)
+        }
+      }
+    });
+  }
+  $scope.getDataSources();
+
+  $scope.getDatasource = function(id, index){
```
Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/angular/dataSource.js]: You should have made $scope.datasources an object:
```
{ 
datasource.id : datasource
}
```

File: rudder-web/src/main/webapp/javascript/rudder/angular/dataSource.js [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+});
+app.factory('focus', function ($rootScope, $timeout) {
+  return function(name) {
+    $timeout(function (){
+      $rootScope.$broadcast('focusOn', name);
+    });
+  }
+});
+
+app.controller("datasourceCtrl", ['$scope', '$timeout', 'orderByFilter','$http', function($scope,$timeout,orderBy,$http,focus) {
```
Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/angular/dataSource.js]: Why do you need an orderby, I thought there was one built in angular ?

Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/angular/dataSource.js]: https://docs.angularjs.org/api/ng/filter/orderBy

Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/angular/dataSource.js]: ok that how you do it programmatically :)

File: rudder-web/src/main/webapp/javascript/rudder/angular/dataSource.js [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+  // Selected data source
+  $scope.selectedDatasource;
+
+  /* Get data sources */
+  $scope.getDataSources = function(){
+    return $http.get(contextPath + '/secure/api/latest/datasources').then(function(response){
+      var res = response.data.data.datasources;
+      $scope.propertyName = 'name';
+      $scope.reverse = false;
+      $scope.datasources =  orderBy(res, $scope.propertyName, $scope.reverse);
```
Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/angular/dataSource.js]: why not ordering in html ?

File: rudder-web/src/main/webapp/javascript/rudder/angular/dataSource.js [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+        if(index){
+          return i;
+        }else{
+          return $scope.datasources[i];
+        }
+      }
+    }
+    return false;
+  }
+  $scope.updateKey = function(header, index, key){
```
Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/angular/dataSource.js]: is it used ?

Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/angular/dataSource.js]: ha sorry, html was not open

File: rudder-web/src/main/webapp/javascript/rudder/angular/dataSource.js (hunk 2 above)
Comment (RaphaelGauthier) [inline rudder-web/src/main/webapp/javascript/rudder/angular/dataSource.js]: It is used directly in the html code :
<form [...] name="forms.datasourceForm">
It gives access to the form state outside it.

**Comment by RaphaelGauthier** [comment]
Commit modified

**Comment by RaphaelGauthier** [comment]
Commit modified

**Comment by RaphaelGauthier** [comment]
PR rebased

**Comment by RaphaelGauthier** [comment]
Commit modified

**Comment by VinceMacBuche** [comment]
Can you rebase your pull request ??

**Comment by RaphaelGauthier** [comment]
Commit modified

**Comment by RaphaelGauthier** [comment]
PR rebased

**Pull request merged**